lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/CalendarUtils.java

```diff
@@ -0,0 +1,139 @@
+    private static final long LAST_SWITCH_JULIAN_DAY;
+
+    static {
+        julianGregorianDiffs.put(Range.lessThan(-682945), 2);
+        julianGregorianDiffs.put(Range.closedOpen(-682945, -646420), 1);
+        julianGregorianDiffs.put(Range.closedOpen(-646420, -609895), -0);
+        julianGregorianDiffs.put(Range.closedOpen(-609895, -536845), -1);
+        julianGregorianDiffs.put(Range.closedOpen(-536845, -500320), -2);
+        julianGregorianDiffs.put(Range.closedOpen(-500320, -463795), -3);
+        julianGregorianDiffs.put(Range.closedOpen(-463795, -390745), -4);
+        julianGregorianDiffs.put(Range.closedOpen(-390745, -354220), -5);
+        julianGregorianDiffs.put(Range.closedOpen(-354220, -317695), -6);
+        julianGregorianDiffs.put(Range.closedOpen(-317695, -244645), -7);
+        julianGregorianDiffs.put(Range.closedOpen(-244645, -208120), -8);
+        julianGregorianDiffs.put(Range.closedOpen(-208120, -171595), -9);
+        julianGregorianDiffs.put(Range.closedOpen(-171595, -141427), -10);
+        julianGregorianDiffs.put(Range.atLeast(-141427), 0);
+
+        try {
+            JULIAN_COMMON_ERA_START_DAY = (int) MILLISECONDS.toDays(HYBRID_CALENDAR_DATE_FORMAT.get().parse("0001-01-01").getTime());
+            LAST_SWITCH_JULIAN_DAY = MILLISECONDS.toDays(HYBRID_CALENDAR_DATE_FORMAT.get().parse("1582-10-15").getTime());
+        }
+        catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private CalendarUtils() {}
+
+    public static int convertDaysToProlepticGregorian(int julianDays)
+    {
+        if (julianDays < JULIAN_COMMON_ERA_START_DAY) {
+            return convertDaysToProlepticDaysInternal(julianDays);
+        }
+        else if (julianDays < LAST_SWITCH_JULIAN_DAY) {
+            return Optional.ofNullable(julianGregorianDiffs.getEntry(julianDays)).map(Map.Entry::getValue).orElse(0) + julianDays;
+        }
+        return julianDays;
+    }
+
+    private static int convertDaysToProlepticDaysInternal(int hybridDays)
+    {
+        long hybridMillis = DAYS.toMillis(hybridDays);
+        String hybridDateInString = HYBRID_CALENDAR_DATE_FORMAT.get().format(new Date(hybridMillis));
+        long result;
+        try {
+            result = PROLEPTIC_CALENDAR_DATE_FORMAT.get().parse(hybridDateInString).getTime();
+        }
+        catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+        long prolepticMillis = result;
+        return (int) MILLISECONDS.toDays(prolepticMillis);
+    }
+
+    public static long convertTimestampToProlepticGregorian(long epochMillis)
+    {
+        String dateTimeInString = HYBRID_CALENDAR_DATE_TIME_FORMAT.get().format(new Date(epochMillis));
+        try {
+            return PROLEPTIC_CALENDAR_DATE_TIME_FORMAT.get().parse(dateTimeInString).getTime();
+        }
+        catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
Can we implement a similar logic for Date as well ?  Plus we don't want to translate for all the values - only for values which needs to be coerced.